File: Export/Interfaces/IPdfExport.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Database;
using System.ComponentModel;

namespace Export
{
    /// <summary>
    /// Interface for implementing PDF exporters.
    /// </summary>
    public interface IPdfExport
    {
        /// <summary>
        /// Name (description) of this exporter
        /// </summary>
        string Name { get; }

        string DocumentTitle { get; set; }
        string DocumentHeader { get; set; }
        string DocumentDate { get; set; }
        string OrgName { get; set; }

        /// <summary>
        /// The list of members to include in this export.
        /// </summary>
        IEnumerable<Member> Members { get; set; }

        /// <summary>
        /// Path to the output file.
        /// </summary>
        string OutputPath { get; set; }

        /// <summary>
        /// Export the job.
        /// </summary>
        void Export();

        /// <summary>
        /// Name (description) of this exporter
        /// </summary>
        string ToString();

        /// <summary>
        /// Happens when progress happens in rendering the output.
        /// </summary>
        event EventHandler<ProgressChangedEventArgs> ProgressChanged;

        /// <summary>
        /// Happens when the export is completed.
        /// </summary>
        event EventHandler ExportCompleted;
    }
}

File: ExportF/IPdfExport.fs
namespace ExportF

open System
open System.Collections.Generic
open System.Linq
open System.Text
open System.ComponentModel

open Database

/// Interface for implementing PDF exporters.
type public IPdfExport = 
    /// Name (description) of this exporter
    abstract member Name : string with get
    abstract member DocumentTitle : string with get
    abstract member DocumentHeader : string with get
    abstract member DocumentDate : string with get
    abstract member OrgName : string with get
    /// The list of members to include in this export.
    abstract Members : IPdfExport -> seq<Member>
    abstract Export : (float -> unit) -> unit
    abstract ToString : string
    /// Path to the output file.
    abstract member OutputPath : string with get
    /// Fired when progress is made in rendering the output.
    abstract ProgressChanged : EventHandler<ProgressChangedEventArgs>
    /// Happens when the export is completed.
    abstract ExportCompleted : EventHandler
